valhalla/sif/dynamiccost.h

```diff
@@ -359,10 +360,13 @@ class DynamicCost {
     // you have forward access for the mode you care about
     // you dont care about what mode has access so long as its forward
     // you dont care about the direction the mode has access to
-    return ((edge->forwardaccess() & access_mask_) ||
-            (ignore_access_ && (edge->forwardaccess() & baldr::kAllAccess)) ||
-            (ignore_oneways_ && (edge->reverseaccess() & access_mask_))) &&
-           (edge->use() != baldr::Use::kConstruction);
+    // AND it is not a construction way
+    // OR if it is a construction and you ignore construction access restrictions
+    return ((((edge->forwardaccess() & access_mask_) ||
+              (ignore_access_ && (edge->forwardaccess() & baldr::kAllAccess)) ||
+              (ignore_oneways_ && (edge->reverseaccess() & access_mask_))) &&
+             (edge->use() != baldr::Use::kConstruction)) ||
+            (ignore_construction_ && (edge->use() == baldr::Use::kConstruction)));
```

Comment:
can we logically pair the two construction parts together and reduce some parentheses, this is very hard to read otherwise. here's how i would do it without asinine autoformatting:

```suggestion
// you have access to the edge
return 
(
  (edge->forwardaccess() & access_mask_) ||
  (ignore_access_ && (edge->forwardaccess() & baldr::kAllAccess)) ||
  (ignore_oneways_ && (edge->reverseaccess() & access_mask_))
) &&
// AND you either dont care if its under construction or its not under construction
(
  ignore_construction_ || edge->use() != baldr::Use::kConstruction
);
```

Reply:
As I pointed out before, that is not the same though. On a road under construction, this first part of the `AND` clause will always be false:

```
(
  (edge->forwardaccess() & access_mask_) ||
  (ignore_access_ && (edge->forwardaccess() & baldr::kAllAccess)) ||
  (ignore_oneways_ && (edge->reverseaccess() & access_mask_))
)
```

If a way has the construction tag, then both `edge->forwardaccess()`  and `edge->reverseaccess()` are `false`. See here in the graph.lua: https://github.com/valhalla/valhalla/blob/master/lua/graph.lua#L1873-L1908

So what I do here is allow access regardless of  `edge->forwardaccess()`  and `edge->reverseaccess()`, but only if it is a construction way and if `ignore_construction` is set.

Reply:
I mean I get what you're saying but it doesn't make sense when you look at what the code previously was. If construction turns off access then why did we even need to check it before? I guess it's possible the code before was just redundant? Also possible I'm just being an idiot :)

Reply:
I didn't think about it that way before. You are right, the check previously was not doing anything as long as the graph.lua stays the same. If we can assume that the graph.lua stays the same then we can skip that check. I believe we can simplify the function to this:

```CPP
  inline virtual bool IsAccessible(const baldr::DirectedEdge* edge) const {
    // you can go on it if:
    // you have forward access for the mode you care about
    // you dont care about what mode has access so long as its forward
    // you dont care about the direction the mode has access to
    return
    // you have access to the edge
    (
      (edge->forwardaccess() & access_mask_) ||
      (ignore_access_ && (edge->forwardaccess() & baldr::kAllAccess)) ||
      (ignore_oneways_ && (edge->reverseaccess() & access_mask_))
    ) ||
    // OR it is under construction but you don't care about that
    (
      ignore_construction_ && edge->use() != baldr::Use::kConstruction
    );
  }
```

What do you think?

Reply:
yeah when i saw the code it made me think that in graph.lua we would turn on access for the appropriate modes of transportation and then this line was saving us from routing on it but if you go over and read graph.lua you find a sound explanation: https://github.com/valhalla/valhalla/blob/b40b23a1b0671fec3648b50a2f3687a9a9701727/lua/graph.lua#L1873-L1877

the problem then is that we are kind of in a broken state because of backwards compatibility. if you want to route on something that is under construction you dont know whether its inteded for cars or pedestrians or bikes or whatever because the access is all turned off. still its the best we can do

Reply:
I suppose the question is if in the future the access to construction roads might be turned on in graph.lua and if we want to safeguard against that case here already.

Reply:
So what's the verdict? I changed it so it doesn't include the redundant construction check anymore, instead it adds the ignore_construction check.

Reply:
sorry i should have been less terse, this looks good to me. what you have is the best we can do without planning for breaking changes to the data compatibility